core/src/fork_selector.rs

```diff
@@ -0,0 +1,288 @@
+use crate::replay_stage_ds::ForkProgress;
+use solana_runtime::bank::Bank;
+use solana_sdk::clock::Slot;
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub struct Fork {
+    pub heaviest_slot: Slot,
+    pub fork_contents: Vec<Slot>,
+}
+
+pub trait ForkSelector {
+    /// extracts fork structure from set of frozen banks,
+    /// ascendants and descendants
+    fn extract_forks(
+        &self,
+        frozen_banks: &HashMap<Slot, Arc<Bank>>,
+        ancestors: &HashMap<u64, HashSet<u64>>,
+        descendants: &HashMap<u64, HashSet<u64>>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> Vec<Fork>;
+
+    /// finds best fork out of vector of fork
+    fn find_best_fork(
+        &self,
+        forks: Vec<Fork>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> Option<Fork>;
+}
+
+#[derive(Default)]
+pub struct OptimalForkSelector {}
```

Comment:
This name should be descriptive of the algorithm, not that its the "best" or "optimal"